Named Plans

seanoshea opened the pull request:
Original issues available at https://github.com/material-motion/material-motion-runtime-objc/issues/81 and https://github.com/material-motion/material-motion-runtime-objc/issues/79

cc @jverkoey @willlarche @rcameron as it looks like you've had some contributions to this repo.

seanoshea committed:
Per issue #81 - A few ideas for adding and removing named plans.

seanoshea commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,36 @@ NS_SWIFT_NAME(Transaction)
+ @param name String identifier for the plan
+ */
+- (void)removePlanNamed:(nonnull NSString *)name
+  NS_SWIFT_NAME(remove(name:));
```
Not sure this is the swifty-ist naming here - definitely open to suggestions on this.

jverkoey replied:
I can see the reasoning that led to this change in the api; we may want to revisit the starmap transaction api to make its desired behavior more clear.

Aside:
Specifically, each "operation" made to a transaction is meant to be eventually committed to the scheduler. A cleaner api for this behavior may have been `append(operation:)` where operation is one of AddPlanOperation, AddNamedPlanOperation, and RemoveNamedPlanOperation, with each of these operations being transferrable to the scheduler so that it can take action on them.

jverkoey replied:
So as for this change: each "remove" operation should eventually result in a performer's removePlan method being invoked so that it can tear down any named state.

seanoshea replied:
https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/c47d5309b615b48bb2b6bf5d925a42dbbd1d4a70 has some thoughts on where this _could_ go, but I'm not sure I've fully grasped the intent of adding a `removePlan` method to the `MDMPlanPerforming` protocol.

Specifically, I'm not sure creating an instance of the performer when removing a plan is correct.

I guess one alternative to this would be to alter the `MDMPlan` protocol to look something like:

```
@protocol MDMPlan <NSObject, NSCopying>

- (nonnull Class)performerClass;

- (nullable id<MDMPerforming>)performer;

@end
```

Otherwise, if we need access to the original performer instance, we could allow access via `MDMPerformerGroup` ?

jverkoey replied:
In https://material-motion.gitbooks.io/material-motion-starmap/content/specifications/runtime/named-plans.html the "Remove-then-add" section explores the expected behavior of addPlan/removePlan for named plans on a given target.

Notably, for a given target there should only be one performer acting on a plan for a given name. The type of performer might change when a new plan is added, so when a named plan is associated with a target we need to give the previously-registered performer an opportunity to release any state associated with the name.

jverkoey replied:
![namedplans](https://cloud.githubusercontent.com/assets/45670/18849269/32489bc8-8400-11e6-8651-d753c3a5e88a.png)

Will iterate on this, but this is a rough visualization of the state change when a new named plan is committed.

jverkoey replied:
Note how the Gesture Performer is asked to remove any plans associated with the name "behavior" (in this case, being draggable). The spring performer is then asked to add the SpringTo plan named "behavior".

seanoshea replied:
Added a `MDMNamedPlanPerforming` protocol which looks like:

```
- (void)addPlan:(nonnull id<MDMPlan>)plan
       withName:(nonnull NSString *)name
  NS_SWIFT_NAME(add(plan:withName:));

- (void)removePlan:(nonnull id<MDMPlan>)plan
          withName:(nonnull NSString *)name
  NS_SWIFT_NAME(remove(plan:withName:));
```

and added the callbacks into `MDMTransaction`.

This change did result in one test failure due to overriding the `isEqual:` method on `MDMTransactionLog` which was needed due to changing the `_namedPlans` to being changed over to `NSMutableSet`.

Can make some changes to fix that test, but wanted to know if this is the correct route to go down.

seanoshea committed:
Per issue #81 - Adding removePlan on the MDMPlanPerforming protocol.

seanoshea committed:
Per issue #81 - Trying out a NamedPlanPerforming protocol.

seanoshea committed:
Per issue #81 - Ensuring the removePlan:withName: method is invoked by
retrieving the transaction log from the _namedPlans.

jverkoey commented on src/MDMPerforming.h:
```diff
@@ -114,3 +114,31 @@ NS_SWIFT_NAME(ComposablePerforming)
+ */
+- (void)removePlan:(nonnull id<MDMPlan>)plan
+          withName:(nonnull NSString *)name
+  NS_SWIFT_NAME(remove(plan:withName:));
```
Did you find that it was necessary/helpful to provide the plan object here?

I think this API might be simplified if we assume that the performer will keep some form of name-associated state and can look up the plan accordingly.

seanoshea replied:
I was kinda looking for consistency between the `add` and `remove` APIs. I think its a fair assumption that if the underlying performer is considering using named plans, then it is probably keeping a reference to them indexed on those same names.

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,37 @@ NS_SWIFT_NAME(Transaction)
+ */
+- (void)removePlanNamed:(nonnull NSString *)name
+             fromTarget:(nonnull id)target
+  NS_SWIFT_NAME(remove(name:from:));
```
`removePlan(named:from:)`

seanoshea replied:
Good catch here. Still wrapping my head around `NS_SWIFT_NAME` and Swift nomenclature in general.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -16,31 +16,108 @@
+      }
+    }
+    [_namedPlans removeObject:retrievedLog];
+    [_logs removeObject:retrievedLog];
```
Ah! I see what was happening here now.

It may help to think about the transaction object as being completely unaware of the scheduler's internals. For example, the transaction object might be on another computer. When we "commit" the transaction to a scheduler on another computer we encode each of the recorded operations (add, addNamed, removeWithName) and send them over the wire. The scheduler then takes action on each of these operations as though these methods were called directly on the scheduler.

The reason we funnel everything through the transaction object — rather than exposing APIs directly on the scheduler — is to ensure that we will architecturally move towards a place where tooling can interact with the runtime over a wire.

jverkoey replied:
More actionable feedback:
- [x] Remove all internal access to the performers from the transaction object.
- [x] Move all performer coordinating to the performerGroup class and execute it on commit.
- [x] Store all transaction logs in the ordered array (we want to ensure that the order in which actions were taken on the transaction is reflected during the commit).

seanoshea replied:
Think I've hit those three points with https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/9f1c50c1de069e3bc53b08e9c9634453692f47a5, https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/06556d80d47f9a823756faedd969733451a9bd4f and https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/ca5516781ceaae56fbe212f8fe3e6d009bbf803c.

One failing test remains:

```
Test Case '-[UnitTests.SchedulerTests testPlansAreCopied]' started.
/Users/sean.oshea/github/material-motion-runtime-objc/tests/unit/SchedulerTests.swift:33: error: -[UnitTests.SchedulerTests testPlansAreCopied] : XCTAssertNotEqual failed: ("<_TtCC9UnitTests14SchedulerTestsP33_53756AB0E4D1E4CAAD92FB270B9314B713ChangeBoolean: 0x608000035fe0>") is equal to ("<_TtCC9UnitTests14SchedulerTestsP33_53756AB0E4D1E4CAAD92FB270B9314B713ChangeBoolean: 0x608000035fe0>") -
```

which is definitely related to me overriding `isEqual:` in `MDMTransactionLog`. Still looking into that.

seanoshea replied:
Looks like I was wrong there in thinking it has something to do with overriding `isEqual:`. https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/69a44e92d4711de03c206bea5cd12d9c90ff0faa should take care of it though.

seanoshea replied:
Been trying a few things in order to try to hit the `removePlanNamed:` callback with some unit tests, but after moving the logic to `MDMPerformerGroup`, I'm having difficulty seeing the callback fire.

I'm assuming that this is mostly due to this `removeObject` call:

```
- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:nil target:target name:name transactionLogType:MDMTransactionLogTypeRemoveNamedPlan];
  [_logs removeObject:log];
}
```

Basically, this ensures that the scheduler will never see any `removePlan` `MDMTransactionLog`? I was mostly basing this code off https://github.com/material-motion/material-motion-runtime-android/blob/develop/library/src/main/java/com/google/android/material/motion/runtime/Transaction.java#L80.

I'm beginning to think that I should be adding the removePlan transaction into the `_logs` collection instead of just removing it?

https://codecov.io/gh/material-motion/material-motion-runtime-objc/compare/2adef827e4077c164e70631d0be1eb4ea8258a2e...29880f066ef3ae7a83a6c4fefd1f87a398bfef91 is the link for the code coverage for these few tests FWIW.

seanoshea replied:
Seeing your comment below, it looks like all logs should be added to the `_logs` collection. Will take a second try at this and update the tests.

seanoshea committed:
Per issue #81 - Code review comments. Renaming the removePlan API.

seanoshea commented on src/MDMPerforming.h:
```diff
@@ -114,3 +114,31 @@ NS_SWIFT_NAME(ComposablePerforming)
+NS_SWIFT_NAME(NamedPlanPerforming)
+@protocol MDMNamedPlanPerforming <MDMPerforming>
+
+@optional
```
In the `Performer Specification` section at https://material-motion.gitbooks.io/material-motion-starmap/content/specifications/runtime/named-plans.html, the spec mentions the following for add/remove operations:

```
If one method is implemented, the other must be implemented as well
```

I'm not sure there's a viable way to enforce this type of rule with protocols in Objective-C, so I made the two methods optional. Perhaps they should both be required?

jverkoey replied:
Marking both methods `required` (the default) on the protocol will require that an object conforming to the protocol implements both methods.

seanoshea replied:
Sure thing. Makes sense. I guess if people aren't interested in implementing both of these methods, then `NamedPlanPerforming` isn't really for them.

seanoshea committed:
Per issue #81 - The spec mentions that if you implement one of these
methods, you should implement the both of them. Marking them as
required. If clients don’t want to implement these methods at all, they
shouldn’t include MDMNamedPlanPerforming in their class signature.

seanoshea committed:
Per issue #81 - Unnecessary checks.

seanoshea committed:
Per issue #81 - Adding an initWithPlan convenience initializer for
MDMTransactionLog.

seanoshea committed:
Per issue #81 - Consolidated to just using the ordered array.

seanoshea committed:
Per issue #81 - Moving this logic to the performer group.

seanoshea committed:
Per issue #81 - Introducing MDMTransactionLogType.

seanoshea committed:
Per issue #81 - Plans should be copied before they are added to the
transaction.

seanoshea committed:
Merge branch 'develop' into issues/81-plan-tests

seanoshea committed:
Per issue #81 - Upgrading tests after sync from develop.

seanoshea committed:
Per issue #81 - Some more tests.

seanoshea committed:
Per issue #81 - A few more tests.

jverkoey commented on src/MDMPerforming.h:
```diff
@@ -148,6 +148,30 @@ NS_SWIFT_NAME(ComposablePerforming)
+@protocol MDMNamedPlanPerforming <MDMPerforming>
+
+/**
+ Provides the performer with an plan and a name.
```
minor: "a plan and associated name"

jverkoey commented on src/MDMPerforming.h:
```diff
@@ -148,6 +148,30 @@ NS_SWIFT_NAME(ComposablePerforming)
+ */
+- (void)addPlan:(nonnull id<MDMPlan>)plan
+       withName:(nonnull NSString *)name
+  NS_SWIFT_NAME(add(plan:withName:));
```
Perhaps `addPlan:named:` would be more consistent with the `removePlanNamed:` api. What do you think?

seanoshea replied:
Hmmmm. Perhaps the `removePlan` function on `MDMPerforming` should be more like `removePlan:withName:` per https://material-motion.gitbooks.io/material-motion-starmap/content/specifications/runtime/named-plans.html.

seanoshea replied:
Thinking of going with `addPlan(_:named:)` and `removePlan(named:)`

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,37 @@ NS_SWIFT_NAME(Transaction)
+Associates a plan with a given target.
+
+@param plan The plan to add to this transaction.
+@param target The target on which the plan can operate.
```
Minor: indent comments by one space.

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,37 @@ NS_SWIFT_NAME(Transaction)
 - (void)addPlan:(nonnull id<MDMPlan>)plan
        toTarget:(nonnull id)target
-    NS_SWIFT_NAME(add(plan:to:));
+       withName:(nonnull NSString *)name
```
`named:`

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,37 @@ NS_SWIFT_NAME(Transaction)
+  NS_SWIFT_NAME(add(plan:to:withName:));
+
+/**
+ Removes a named plan from the transaction.
```
`Remove any plan associated with given name on the given target.`

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,37 @@ NS_SWIFT_NAME(Transaction)
+ 
+ @param plan The plan to add to this transaction.
+ @param target The target on which the plan can operate.
+ @param name String identifier for the plan
```
minor: Add missing . to end of `@param name`.

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,37 @@ NS_SWIFT_NAME(Transaction)
+/**
+ Removes a named plan from the transaction.
+ 
+ @param plan The plan to add to this transaction.
```
Remove docs for nonexistent `plan` parameter.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,65 @@ - (instancetype)init {
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLogType transactionLogType = name == nil ? MDMTransactionLogTypeUncategorized : MDMTransactionLogTypeAddNamedPlan;
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:[plan copy] target:target name:name transactionLogType:transactionLogType];
```
Our API guarantees that `name` will be non-null when using the named plan APIs.

Can we use the null state of the `name` to identify whether this is a named plan operation or a regular plan operation? This might allow us to remove the `MDMTransactionLogType` enum.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,65 @@ - (instancetype)init {
+  return self;
+}
+
+- (BOOL)isEqual:(id)other {
```
We likely don't need this method implementation if we ensure that all of the operations on the transaction are stored in the transaction log and then committed to the performer group.

seanoshea replied:
Yup - looks like me overriding `isEqual:` was overkill.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,65 @@ - (instancetype)init {
 
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:nil target:target name:name transactionLogType:MDMTransactionLogTypeRemoveNamedPlan];
+  [_logs removeObject:log];
```
This should be `_logs addObject:log`, or our performer group will never receive the operation.

jverkoey commented on src/private/MDMTransaction+Private.h:
```diff
@@ -18,9 +18,22 @@
+@property(nonatomic, copy) NSArray<id<MDMPlan>> *plans;
+@property(nonatomic, strong) id target;
+@property(nonatomic, copy) NSString *name;
+@property(nonatomic, assign) MDMTransactionLogType transactionLogType;
```
This might be representable as `BOOL isRemoval` if we use nullable state of `name` to indicate whether this is a named operation.

seanoshea committed:
Per issue #81 - Code review comments. Still a WIP.

seanoshea committed:
Per issue #81 - Compilation issues.

seanoshea commented:
Even if all the checks have passed, please do not merge this as there are two failing unit tests still that I'm in the process of figuring out. One is:
```
  func testAddingTheSameNamedPlanTwiceToTheSameTarget() {
    let firstPlan = IncrementerTargetPlan()
    transaction.add(plan: firstPlan, to: incrementerTarget, withName: "one")
    transaction.add(plan: firstPlan, to: incrementerTarget, withName: "one")
    
    scheduler.commit(transaction: transaction)
    
    XCTAssertTrue(incrementerTarget.counter == 1)
  }
```
Note: the `removePlan:named` callback is now being invoked, but I'm still not actually removing the correct `MDMPlan` from the `MDMTrace`'s `committedPlans`.

Also, it looks like we need a `NamedPlan` type of `MDMPlan` per https://material-motion.gitbooks.io/material-motion-starmap/content/specifications/runtime/named-plans.html, so I have to add that still.

seanoshea committed:
Per issue #81 - WIP One remaining failing unit test.

jverkoey commented on src/MDMPerforming.h:
```diff
@@ -148,6 +148,30 @@ NS_SWIFT_NAME(ComposablePerforming)
+ */
+- (void)addPlan:(nonnull id<MDMPlan>)plan
+          named:(nonnull NSString *)name
+  NS_SWIFT_NAME(addPlan(plan:named:));
```
Hm, this API brings up a good point. I agree here that `addPlan` is more consistent with `removePlan`, so in order to avoid the redundancy of `addPlan(plan:named:)` we can make the first argument non-named: `addPlan(_:named:)`.

This will allow us to invoke the method like so: `addPlan(SomePlan, named:"name")` rather than `addPlan(plan:SomePlan, named:"name")`.

jverkoey replied:
Filed https://github.com/material-motion/material-motion-runtime-objc/issues/89 to track changes to existing `add(plan:)` non-named methods.

seanoshea replied:
Ah - I was futzing around with `NS_SWIFT_NAME` forever trying to figure out how to have an unnamed param. Thanks for the pointer.

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -27,9 +27,36 @@ NS_SWIFT_NAME(Transaction)
        toTarget:(nonnull id)target
-    NS_SWIFT_NAME(add(plan:to:));
+          named:(nonnull NSString *)name
+  NS_SWIFT_NAME(add(plan:to:withName:));
```
Similarly here, I think `addPlan(_:to:named:)` may be more consistent with our `removePlan` API.

jverkoey replied:
Am appreciating that Swift API naming is encouraging thinking about these as sentences. In that manner "add this plan named 'interaction' to this target" is probably less ambiguous than "add this plan to this target named 'interaction'". In the latter case I can imagine confusion by assuming that the target is the thing being named.

Perhaps `addPlan:named:toTarget:` and `addPlan(_:named:to:)` will be cleaner overall? What do you think? I'll update the spec if we choose to go this route.

seanoshea replied:
I _think_ `addPlan:named:toTarget` makes most sense to me. If I look at `addPlan:toTarget:withName`, I'm a little confused as to whether the `name` relates to the name of the target or the name of the plan.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target named:(NSString *)name {
+  if (name.length && [self planForName:name withTarget:target]) {
```
I see you're special-casing the empty string here; can I learn more about your rationale here?

If we do decide to go this route we'll need to document the behavior, which leads me to wonder if we should be special-casing the behavior or simply allowing empty strings as names.

seanoshea replied:
Yeah - I think a check against the empty string would be reasonable here. I don't think we'd want to allow named plans to be added without names. Something like:

```
NSParameterAssert(name.length > 0);
```

might be reasonable?

seanoshea replied:
There's two places where this PR checks for `name.length > 0`. Could be worthwhile to extract this into a method on `MDMTransactionLog` if you think it'd be useful.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
+}
+
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  [self commonRemovePlan:[self planForName:name withTarget:target] fromTarget:target named:name];
```
`[self planForName:name withTarget:target]` is building more assumptions into the Transaction object than it needs to be responsible for.

A Transaction is often an ephemeral instance created at any time and containing any number of requested operations. A Transaction might look like this:

```
let transaction = Transaction()
transaction.removePlan(named:"someName")
scheduler.commit(transaction: transaction)
```

In this scenario the transaction understandably wouldn't know which plan is associated with a given name. The code as written has an expectation that the transaction is aware of plans registered to a given target, rather than simply relaying "remove a plan with a given name" to the scheduler.

To emphasize previous discussions about the transaction object: it is meant to be a very dumb vessel that captures requested operations and forwards those operations to the scheduler. The transaction has no long-term awareness of ongoing state; it solely enables communication from the application code to the scheduler object.
- [x] The Transaction object's public APIs should simply be logged as "operations" in an ordered array, without any form of optimization or other state manipulation.
- [x] When a transaction is committed to the scheduler, the scheduler enumerates each of these operations as though they were invoked on the scheduler itself.

seanoshea replied:
Thanks for the explanations here. Makes sense to move all this logic into the PerformerGroup.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
   return _logs;
 }
 
+- (id<MDMPlan>)planForName:(NSString *)name withTarget:(id)target {
```
- [x] Delete this method. It will encourage too much state manipulation within the transaction object.

seanoshea replied:
Yup - nuked this method.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,17 +32,58 @@ - (instancetype)init {
+- (instancetype)initWithPlan:(NSObject<MDMPlan> *)plan target:(id)target name:(NSString *)name removal:(BOOL)removal {
+  self = [super init];
+  if (self) {
+    if (plan != nil) {
```
Plan should never be allowed to be nil here due to nullability APIs. We can throw an assertion to be safe.

seanoshea replied:
The reason I was checking for `nil` here was because I was using this convenience initializer for creating `MDMPlans` from `commonRemovePlan` where we don't have a reference to a plan when we're trying to remove it. I guess I could use a different initializer for the remove case?

Given the fact that we want to keep the transaction object as simple as possible, I'm not sure I should be trying to look up the a plan associated with a name and a target ala `planForName:withTarget:`, hence the nil check.

seanoshea replied:
Happy to either open up a different initializer strictly for `removePlanNamed:fromTarget` if you think the `nil` check here is a little odd.

jverkoey replied:
I think providing a separate initializer for each type of operation would be reasonable and possibly the "cleanest" API, in that it doesn't allow any ambiguity of usage.

jverkoey commented on src/private/MDMPerformerGroup.m:
```diff
@@ -72,6 +72,19 @@ - (void)executeLog:(MDMTransactionLog *)log trace:(MDMTrace *)trace {
     }
+    
+    // MDMNamedPlanPerforming callbacks
+    if (log.name.length) {
```
Assuming we support empty names, presence of a non-nil `name` property will be a sufficient check here.

seanoshea replied:
https://material-motion.gitbooks.io/material-motion-starmap/content/specifications/runtime/named-plans.html indicates that the name should not be null or the empty string.

Not sure I can envisage a reasonable use case for an named plan which has an empty name regardless. Would we not steer people towards the standard `addPlan` method if they didn't want to include a name?

jverkoey commented on src/private/MDMPerformerGroup.m:
```diff
@@ -72,6 +72,19 @@ - (void)executeLog:(MDMTransactionLog *)log trace:(MDMTrace *)trace {
+        if ([performer respondsToSelector:@selector(addPlan:named:)]) {
+          [(id<MDMNamedPlanPerforming>)performer addPlan:plan named:log.name];
+        }
+      }
```
The pseudo-code for adding/removing named plans in the performer group should look roughly like so:

```
# Step 1
performerForName(name).removePlan(named: name)

# Step 2
performer = performerForPlan(plan)
performer.add(plan: plan, withName: name)
performerForName(name) == performer 
> true
```

Remove-then-add: Two things happen when a named plan is committed:
1. Remove any previously committed plan with the same name from the target's performers.
   Note: This may be on a different performer instance on the same target. In the example above perhaps a PhysicsPerformer is needed for the second transaction, but not for the first.
2. Provide the relevant performer with the new named plan.

seanoshea replied:
I don't strictly have a `performerForName` method in `executeLog` method in `MDMPerformerGroup`, but the existing `performerForPlan:isNew:` method seems to do the trick.
Would there be any scenario whereby a performed associated with a named plan would need to be looked up by name instead of by it's plan's performer?

jverkoey replied:
We'll find it beneficial to keep track of which performer was associated with a given name. Otherwise, each time a named plan is added we'd have to ask every performer to remove any plan associated with `name` because we wouldn't know who has one.

So we need both types of lookups. performerForPlan instantiates the performer for us, performerForName tells us which performer was previously the "owner" of that name.

seanoshea replied:
Yup - makes sense with that explanation. Will take a pop at integrating this.

seanoshea committed:
Per issue #81 - Adding NamedPlan as a protocol. Code review feedback
too. Removing logic from the transaction and housing it in the
performer group instead.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,10 +32,25 @@ - (instancetype)init {
+}
+
+- (void)commonRemovePlan:(NSObject<MDMPlan> *)plan fromTarget:(nonnull id)target named:(NSString *)named {
+  [_logs addObject:[[MDMTransactionLog alloc] initWithPlan:plan target:target name:named removal:TRUE]];
```
Perhaps we should move the `[plan copy]` code into the MDMTransactionLog's initializer? Need to enforce the fact that plans are copied into a transaction the moment they're added.

jverkoey commented on src/MDMTransaction.m:
```diff
@@ -30,10 +32,25 @@ - (instancetype)init {
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan named:(NSString *)name toTarget:(id)target {
+  NSParameterAssert(name.length > 0);
+  [self commonRemovePlan:plan fromTarget:target named:name];
```
I think we can nuke this extra `removePlan` log here because the performer group should be responsible for removing previous named plans whenever a named plan is added.

Consider when we expose the `addPlan:named:` method on the scheduler: we can't expect clients to remember to call removePlan each time.

seanoshea replied:
Sure thing - will find a home for this in `executeLog:` in the performer group.

seanoshea committed:
Per issue #81 - Moving more logic to MDMPerformerGroup.

seanoshea committed:
Per issue #81 - Merging with the latest from develop.

seanoshea commented:
Rebased from develop and I think I have a solution which passes all previous unit tests. I think there's a few different ways to proceed:
1. Remove the `addPlan:named:toTarget` and `removePlanNamed:fromTarget:` from `MDMTransaction` and add them to `MDMScheduler` instead.
2. Continue as is with these methods living on `MDMTransaction`. Remove them at a later date.
I'd tend to go for the first option as I think it's better to get rid of these methods before they are part of any public API. Definitely open to suggestions though if you think otherwise.

seanoshea committed:
Per issue #81 - Moving logic into the scheduler.

seanoshea committed:
Per issue #81 - Adding a performerPlanNameToPerformerInfo lookup in the
PerformerGroup. Moving some removal logic from the scheduler to the
PerformerGroup too.

seanoshea commented on src/private/MDMPerformerGroup.m:
```diff
@@ -76,30 +78,64 @@ - (void)terminateIsActiveToken:(id<MDMIsActiveTokenable>)token
+    if (plan != nil) {
+      [trace.committedPlans addObject:plan];
+    } else {
+      // this is the case whereby we are calling removePlan:named, but don't have a MDMPlan
```
I think I'm being a little inconsistent here when adding/removing plans. Consider the following two cases:

**Case 1 - A very simple add named plan**

```
scheduler.addPlan(firstViewTargetAlteringPlan, named: "name_one", to: target)
// The callbacks would happen like `removePlanNamed:`, `addPlan:named`
// the trace.committedPlans would contain a removePlan, then an addPlan
```

**Case 2 - A very simple add named plan followed by a remove named plan**

```
scheduler.addPlan(firstViewTargetAlteringPlan, named: "name_one", to: target)
scheduler.removePlan(named: "name_one", from: target)
// The callbacks would happen like `removePlanNamed:`, `addPlan:named`, `removePlanNamed:`
// the trace.committedPlans would contain a removePlan, then an addPlan.
```

As `trace.committedPlans` is internal and not really accessible by the client, I'm unsure of whether this is an issue. However, the Notifications sent out might be a little confusing in this case. Basically, this unit test fails, but I'm not sure it should:

```
  func testAddAndRemoveTheSameNamedPlan() {
    let scheduler = Scheduler()

    var numberOfPlansCommitted = 0
    expectation(forNotification: TraceNotificationName.plansCommitted._rawValue as String, object: scheduler) { notification -> Bool in
      let event = notification.userInfo![TraceNotificationPayloadKey] as! SchedulerPlansCommittedTracePayload
      numberOfPlansCommitted = numberOfPlansCommitted + event.committedPlans.count
      // I might expect this to be true given that the callbacks executed were
      // `removePlanNamed:`, `addPlan:named`, `removePlanNamed:`
      // how could a callback get invoked without a plan being committed?
      return numberOfPlansCommitted == 3
    }

    scheduler.addPlan(firstViewTargetAlteringPlan, named: "name_one", to: target)
    scheduler.removePlan(named: "name_one", from: target)

    XCTAssertTrue(target.text! == "removePlanInvokedaddPlanInvokedremovePlanInvoked")

    waitForExpectations(timeout: 0.1)
  }
```

I guess we could get around this issue by storing the `name` and `target` properties on `MDMNamedPlan` and whenever we get a `removePlan:named` operation in this `else` clause, try look back through the `trace.committedPlans` for a named plan which matches the `name` and `target` that the client just provided in the `removePlanNamed:fromTarget` call in the Scheduler. Not sure if this "solution" is very elegant & unsure of whether we'd want to add these two properties to `MDMNamedPlan`

The other way around this would be to allow `MDMPerformerGroup` to keep a reference of all the committed `MDMTransactionLog` objects?

jverkoey replied:
We may want two different storage units for plans in the tracer in that case, one for `committedPlans` and another for `removedPlans`.

I've found the trace notifications approach to be moderately challenging to work with, so I have a diff out that simplifies the system: http://codereview.cc/D1682. If we go with that approach instead then you can add a new method to the Tracer type for `didAddPlan(_:named:to:)`. This'll make writing the unit tests somewhat simpler.

seanoshea replied:
OK - will take a look at that code review.

Need to figure out exactly what data structure would be in the `removedPlans` collection as we don't always have a `MDMPlan` object when calling `removePlan:named:from`. I'm guessing `removedPlans` would contain a structure of names associated with targets.

or

Whenever we get a `removePlan:named:from` committed to the scheduler, we should look back through `committedPlans` and see if we can find a `MDMNamedPlan` which was previously added which matches the name and target. If we find a plan, copy it over into `removedPlans`

seanoshea replied:
I see that `Tracer` PR (http://codereview.cc/D1682) landed with the `didAddPlan(_:named:to:)` method. As this PR includes logic for removing named plans, would you suggest a similar `didRemovePlanNamed(_:from:)` `Tracer` optional method?

seanoshea committed:
Per issue #81 - Separate collections for add and remove plans.

seanoshea commented:
Separated out the `NamedPlans` into `committedAddPlans` and `committedRemovePlans`.

Note that a `removePlan:named:from:` operation does not add a `NamedPlan` to the `committedRemovePlans` collection. Not sure if this is an issue, but thought it worthwhile to point out all the same.

seanoshea committed:
Per issue #81 - README updates associated with named plans.

jverkoey commented on src/MDMScheduler.h:
```diff
@@ -63,6 +64,28 @@ NS_SWIFT_NAME(Scheduler)
+ */
+- (void)addPlan:(nonnull id<MDMNamedPlan>)plan
+          named:(nonnull NSString *)name
+       toTarget:(nonnull id)target
```
`addPlan:named:to:` in Objective-C to keep consistent with the Swift API.

jverkoey commented on src/MDMScheduler.h:
```diff
@@ -63,6 +64,28 @@ NS_SWIFT_NAME(Scheduler)
+ */
+- (void)removePlanNamed:(nonnull NSString *)name
+             fromTarget:(nonnull id)target
+NS_SWIFT_NAME(removePlan(named:from:));
```
`removePlanNamed:from:` in Objective-C to keep consistent with the Swift API.

jverkoey commented on src/private/MDMPerformerGroup.h:
```diff
@@ -37,7 +37,7 @@
 @property(nonatomic, nullable) id schedulerTarget;
 
-- (void)addPlan:(nonnull id<MDMPlan>)plan trace:(nonnull MDMTrace *)trace;
+- (void)addPlan:(nonnull id<MDMPlan>)plan trace:(nonnull MDMTrace *)trace log:(nullable MDMTransactionLog *)log;
```
You may find this implementation to be cleaner to implement if you break it out to match the public APIs of addPlan, addPlanNamed, and removePlanNamed and then drop use of the transaction log object. Had you considered this already?

seanoshea replied:
Hmmm. I guess I was mostly using `- (void)addPlan:(nonnull id<MDMPlan>)plan trace:(nonnull MDMTrace *)trace log:(nullable MDMTransactionLog *)log` for _everything_ and then twisting and turning within that method with if checks for each of the different variants. I'll have a look into breaking this out.

seanoshea replied:
Added the following methods to `MDMPerformerGroup`:

```
- (void)addPlan:(nonnull id<MDMPlan>)plan to:(nonnull id)target trace:(nonnull MDMTrace *)trace;
- (void)addPlan:(nonnull id<MDMNamedPlan>)plan named:(nonnull NSString *)name to:(nonnull id)target trace:(nonnull MDMTrace *)trace;
- (void)removePlanNamed:(nonnull NSString *)name from:(nonnull id)target;
```

which is helping to split up this logic a little.

seanoshea committed:
Per issue #81 - Merge latest develop.

seanoshea committed:
Per issue #81 - Code review comments. Bringing objective-c signatures
inline with swift.

seanoshea committed:
Per issue #81 - Using the correct plan.

seanoshea committed:
Merge branch 'develop' into issues/81-plan-tests

pingpongboss commented on src/MDMPlan.h:
```diff
@@ -35,3 +35,12 @@ NS_SWIFT_NAME(Plan)
+
+/**
+ Instances of `MDMNamedPlan` can be added or removed as named plans to or from targets.
+
```
extra whitespace

pingpongboss commented on src/MDMScheduler.m:
```diff
@@ -81,16 +81,33 @@ - (MDMSchedulerActivityState)activityState {
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan named:(NSString *)name to:(id)target {
+  NSParameterAssert(name.length > 0);
+  [self commonAddPlan:plan toTarget:target log:[[MDMTransactionLog alloc] initWithPlans:@[plan] target:target name:name removal:FALSE]];
```
i find the `removal:FALSE` init param a little clunky. Especially since the `removal:TRUE` case is just omitted

seanoshea replied:
Created a separate initializer in https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/15282b9b3d66bc8d71c9bbe4893cadbbf747726b & https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/b2639f4a41293429680639ccc50c9180b7743aeb

pingpongboss commented on src/MDMScheduler.m:
```diff
@@ -81,16 +81,33 @@ - (MDMSchedulerActivityState)activityState {
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan named:(NSString *)name to:(id)target {
+  NSParameterAssert(name.length > 0);
+  [self commonAddPlan:plan toTarget:target log:[[MDMTransactionLog alloc] initWithPlans:@[plan] target:target name:name removal:FALSE]];
```
@jverkoey is "transaction log" as a concept still useful? I find it to be a past vestige that we can now remove. You'd still need a data holder to hold the Plan's metadata (name mostly). But it's pretty redundant to have both `addPlan:plan` and `TransactionLog initWithPlan:plan`

seanoshea replied:
Hmmmmm - I think there's a `PlanInfo` class in the Android codebase which acts as a data holder?

I was wondering whether we could just move all this metadata into the `MDMPlan` itself instead of keeping track of a plan and metadata associated with a plan in two separate objects? Maybe that's a bit of a drastic change?

pingpongboss replied:
Interesting. I think the reason I would hesitate to do that is because Plan is a client facing public object that the client may keep around. And these metadata are private and internal and I'd rather we keep track of it internally.

seanoshea replied:
OK. With https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/15282b9b3d66bc8d71c9bbe4893cadbbf747726b & https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/b2639f4a41293429680639ccc50c9180b7743aeb, I've taken any mentions of `MDMTransactionLog`out of `MDMScheduler`. The only remaining references relate to deprecated methods now.

pingpongboss commented on src/MDMScheduler.m:
```diff
@@ -81,16 +81,33 @@ - (MDMSchedulerActivityState)activityState {
+
+- (void)removePlanNamed:(NSString *)name from:(id)target {
+  NSParameterAssert(name.length > 0);
+  [self commonAddPlan:nil toTarget:target log:[[MDMTransactionLog alloc] initWithTarget:target name:name]];
```
It may be a bad abstraction to use `commandAddPlan:nil` for removing a plan.

seanoshea replied:
Yep - refactoring this a little. Should be taken care of in https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/15282b9b3d66bc8d71c9bbe4893cadbbf747726b & https://github.com/material-motion/material-motion-runtime-objc/pull/88/commits/b2639f4a41293429680639ccc50c9180b7743aeb

seanoshea committed:
Per issue #81 - Adding addPlan/removePlan methods to the performer
group. Removing transactions references from the scheduler.

seanoshea committed:
Per issue #81 - Consolidation.

seanoshea committed:
Per issue #81 - Doesn’t need to be a separate function.

jverkoey commented on Podfile.lock:
```diff
@@ -1,6 +1,6 @@
 PODS:
   - CatalogByConvention (1.0.1)
-  - MaterialMotionRuntime (3.0.1)
+  - MaterialMotionRuntime (3.0.0)
```
Revert this change.

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -29,11 +30,16 @@ NS_SWIFT_NAME(Transaction)
-    NS_SWIFT_NAME(add(plan:to:))
-    __deprecated_msg("Add plans directly to a scheduler instead.");
+  NS_SWIFT_NAME(add(plan:to:))
+  __deprecated_msg("Add plans directly to a scheduler instead.");
```
Revert these changes.

jverkoey commented on src/MDMTransaction.h:
```diff
@@ -17,6 +17,7 @@
 #import <Foundation/Foundation.h>
 
 @protocol MDMPlan;
+@protocol MDMNamedPlan;
```
Revert this change.

jverkoey commented on src/private/MDMPerformerGroup.m:
```diff
@@ -93,28 +100,55 @@ - (void)terminateIsActiveToken:(id<MDMIsActiveTokenable>)token
+    }
+  } else {
+    // this is the case whereby the client has tried to remove a named performer which was never added in the first place
+  }
```
We can simplify this implementation by leaning on nil-messaging behavior.
- [x] Remove the if/else statement.

jverkoey commented on src/private/MDMPerformerGroup.m:
```diff
@@ -42,35 +43,41 @@ - (instancetype)initWithTarget:(id)target scheduler:(MDMScheduler *)scheduler {
+}
+
+- (void)removePlanNamed:(nonnull NSString *)name from:(nonnull id)target {
+  [self removePlanNamed:name from:target withPerformerInfo:self.performerPlanNameToPerformerInfo[name]];
```
We don't appear to be removing the performer info from `performerPlanNameToPerformerInfo` after we remove the named plan.
- [ ] Remove any associated performerInfo from the given name in removePlanNamed...

seanoshea replied:
Still looking into this, but removing the performer in `removePlanNamed` results in this unit test failing:

```
func testAddAndRemoveTheSameNamedPlan() {
    let scheduler = Scheduler()

    scheduler.addPlan(firstViewTargetAlteringPlan, named: "name_one", to: target)
    scheduler.removePlan(named: "name_one", from: target)

    XCTAssertTrue(target.text! == "removePlanInvokedaddPlanInvokedremovePlanInvoked")
  }
```

jverkoey commented on src/private/MDMPerformerGroup.m:
```diff
@@ -93,28 +100,55 @@ - (void)terminateIsActiveToken:(id<MDMIsActiveTokenable>)token
-    return performerInfo.performer;
+    return performerInfo;
+  } else {
+    id<MDMPerforming> performer = [[performerClass alloc] initWithTarget:self.target];
```
- [x] Remove the `} else {` clause here so that we have one less layer of scope. Don't remove the contents of the else clause. We lean on the fact that the if statement `returns`.

jverkoey commented on src/private/MDMTransaction+Private.h:
```diff
@@ -19,8 +19,18 @@
+
+- (instancetype)initWithPlans:(NSArray<NSObject<MDMPlan> *> *)plans target:(id)target name:(NSString *)name;
+
+- (instancetype)initWithTarget:(id)target name:(NSString *)name;
```
- [ ] If possible, revert these changes so as not to add complexity to this commit.

pingpongboss commented on src/MDMTransaction.m:
```diff
@@ -43,4 +42,36 @@ - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
 
 @implementation MDMTransactionLog
+
+- (instancetype)initWithPlans:(NSArray<NSObject<MDMPlan> *> *)plans target:(id)target name:(NSString *)name removal:(BOOL)removal {
```
who calls this method with the `removal` param?

seanoshea replied:
Nobody anymore by the looks of things. I'll take it out.

pingpongboss commented on src/private/MDMTransaction+Private.h:
```diff
@@ -19,8 +19,18 @@
+
+@property(nonatomic, copy) NSArray<id<MDMPlan>> *plans;
+@property(nonatomic, strong) id target;
+@property(nonatomic, copy) NSString *name;
```
I'm not sure if it makes sense anymore to have a `plans` array _and_ a `name` field. Since multiple plans cannot share the same name (right @jverkoey ?) Not sure if this is in scope for this PR

pingpongboss commented on src/private/MDMPerformerGroup.m:
```diff
@@ -42,35 +43,41 @@ - (instancetype)initWithTarget:(id)target scheduler:(MDMScheduler *)scheduler {
+  BOOL isNew = NO;
+  MDMPerformerInfo *performerInfo = [self findOrCreatePerformerInfoForNamedPlan:plan named:name trace:trace isNew:&isNew];
+  [self removePlanNamed:name from:target withPerformerInfo:performerInfo];
+  [trace.committedRemovePlans addObject:plan];
```
should the param here really be `plan`?

seanoshea replied:
An argument could be made that the `trace.committedRemovePlans` should really be the `name`s instead of the plans themselves? Is that what you're referring to?

pingpongboss replied:
Ah what I meant is, `plan` is the "added plan" while it reads like you want to pass in the "removed plan" here. So maybe:

```
Plan removed = [self removePlanNamed:...];
[trace.committedRemovePlans addObject:removed];
```

But it's possible I'm not reading things right.

pingpongboss commented on src/private/MDMPerformerGroup.m:
```diff
@@ -93,28 +100,55 @@ - (void)terminateIsActiveToken:(id<MDMIsActiveTokenable>)token
 #pragma mark - Private
 
-- (id<MDMPerforming>)performerForPlan:(id<MDMPlan>)plan isNew:(BOOL *)isNew {
+- (void)removePlanNamed:(nonnull NSString *)name from:(nonnull id)target withPerformerInfo:(nullable MDMPerformerInfo *)performerInfo {
```
should we be sending out a trace notification?

jverkoey replied:
We can do so in a follow-up diff/PR.

seanoshea replied:
Yeah - I wasn't sure whether this was outside the scope of this PR?

jverkoey replied:
Let's file a follow-up issue and punt.

seanoshea replied:
Opened https://github.com/material-motion/material-motion-runtime-objc/issues/94

seanoshea committed:
Per #81 - Code review comments.

seanoshea committed:
Per issue #81 - Adding back this test.

jverkoey reviewed (approved):
LGTM. There are a few follow-up things to tackle but I think we can land this and iterate.

seanoshea committed:
Per issue #81 - Removing the remove flag.
